lib/FusionInventory/Agent/Task/Inventory/Linux/Storages.pm

```diff
@@ -56,24 +65,29 @@ sub _getDevices {
         }
     }
 
-    # get serial & firmware numbers from hdparm, if available
-    if (_correctHdparmAvailable(%params)) {
-        foreach my $device (@devices) {
-            next if $device->{SERIALNUMBER} && $device->{FIRMWARE};
-            my $info = getHdparmInfo(
-                device => "/dev/" . $device->{NAME},
-                %params
+    # get missing fields using functions defined in %map
+    for my $device (@devices) {
+        # the hash keys are function references from %map
+        my %info;
+
+        OUTER: for my $field (keys %map) {
+            next unless (
+                ($field eq 'MANUFACTURER'
+                 && defined $device->{$field}
+                 && $device->{$field} eq 'ATA')
+                      ||
+                !$device->{$field}
             );
```

Comment:
Also hard to read, if I understand well, I think the following does what you want:
```suggestion
            next if defined($device->{$field}) && $field eq 'MANUFACTURER' ?
                $device->{$field} ne 'ATA' : length($device->{$field});
```
Don't forget `!$device->{$field}` while be true if $device->{$field} is 0 and this may be the case for DISKSIZE. To support the case where a field is defined and empty, I just check its length when defined.

Reply:
Whether it's hard to read or not depends on how much you like ternary operators ;)

Reply:
There is a new version, take a look.